### ISSUE TITLE
System.FormatException: 'Unexpected token 

### ISSUE THREAD

**Issue by ghost1372** [dotnet/wpf#2132]
* .NET Core Version: 3
* Windows version: 18362.499
* Does the bug reproduce also in WPF for .NET Framework 4.8?: Yes
 
 **Problem description:**
 i want to create geometry with `geometry.Parse(mystring);` but i get following error

> System.FormatException: 'Unexpected token 

after some research i found that i can fix problem with this code in app.cs

`
 Thread.CurrentThread.CurrentCulture = CultureInfo.InvariantCulture; 
`

now i have another problem and some of my controls like datepicker not work and get exception

**Comment by miloush** [comment]
Include the string you are trying to parse and details of the exception you are getting after switching the thread culture. What other controls do not work?

How do you reproduce the error? System.Windows.Media.Geometry.Parse explicitly uses invariant culture: https://referencesource.microsoft.com/#PresentationCore/Core/CSharp/System/Windows/Media/Generated/Geometry.cs,321

**Comment by ghost1372** [comment]
@miloush this string is generated by this code
`
M41/416254,90C41/88876,82/5298 41/624584,57/2716 41/416254,50/99 36/29352,49/8907 36/894651,50/4264 31/567267,50/99 31/775597,44/59305 31/931627,42/43234 31/693536,36/77758 36/340008,36/79568 35/133525,36/76858 41/542523,36/58818 41/638383,32/81458 41/675605,33/50908 41/668792,29/20144 41/851382,25/46201 41/182183,18/65836 45/961941,14/23856 50/741699,9/81876 59/568752,10/59048 68/185297,10/70302 68/141127,16/44056 68/145937,20/8288 68/185297,23/58247 63/135373,24/4635 60/876109,23/24558 57/326157,24/84516 55/922779,28/231 55/756556,29/65047 56/63466,36/77758 59/21143,36/64422 64/265254,36/63978 68/690373,36/77758 67/695215,42/78657 67/742088,44/40134 66/922606,49/8464 63/245981,49/9344 61/316885,49/6976 55/810928,49/8464 55/662114,56/6024 56/168075,82/8571 55/810928,90 49/382352,90/1786 47/63649,89/97 41/416254,90z
`

https://github.com/ghost1372/HandyControls/blob/f24752980399c9ae1d749b877b05fd338ef74add/src/Shared/HandyControl_Shared/Tools/Helper/AnimationHelper.cs#L54-L68

exception is

`
System.FormatException: 'Unexpected token 'M41/416254,90C41/88876,82/5298 41/624584,57/2716 41/416254,50/99 36/29352,49/8907 36/894651,50/4264 31/567267,50/99 31/775597,44/59305 31/931627,42/43234 31/693536,36/77758 36/340008,36/79568 35/133525,36/76858 41/542523,36/58818 41/638383,32/81458 41/675605,33/50908 41/668792,29/20144 41/851382,25/46201 41/182183,18/65836 45/961941,14/23856 50/741699,9/81876 59/568752,10/59048 68/185297,10/70302 68/141127,16/44056 68/145937,20/8288 68/185297,23/58247 63/135373,24/4635 60/876109,23/24558 57/326157,24/84516 55/922779,28/231 55/756556,29/65047 56/63466,36/77758 59/21143,36/64422 64/265254,36/63978 68/690373,36/77758 67/695215,42/78657 67/742088,44/40134 66/922606,49/8464 63/245981,49/9344 61/316885,49/6976 55/810928,49/8464 55/662114,56/6024 56/168075,82/8571 55/810928,90 49/382352,90/1786 47/63649,89/97 41/416254,90z' encountered at position '2'.'
`

and when i use invariantCulture string is

`
M41.416254,90C41.088876,82.5298 41.624584,57.2716 41.416254,50.099 36.029352,49.8907 36.894651,50.4264 31.567267,50.099 31.775597,44.59305 31.931627,42.43234 31.693536,36.77758 36.340008,36.79568 35.133525,36.76858 41.542523,36.58818 41.638383,32.81458 41.675605,33.50908 41.668792,29.20144 41.851382,25.46201 41.182183,18.65836 45.961941,14.23856 50.741699,9.81876 59.568752,10.59048 68.185297,10.70302 68.141127,16.44056 68.145937,20.08288 68.185297,23.58247 63.135373,24.04635 60.876109,23.24558 57.326157,24.84516 55.922779,28.0231 55.756556,29.65047 56.063466,36.77758 59.21143,36.64422 64.265254,36.63978 68.690373,36.77758 67.695215,42.78657 67.742088,44.40134 66.922606,49.8464 63.245981,49.9344 61.316885,49.6976 55.810928,49.8464 55.662114,56.6024 56.168075,82.8571 55.810928,90 49.382352,90.1786 47.63649,89.97 41.416254,90z
`

when i use invariantCulture i cant use DatePicker control and i get FormatError Exception that year is incorrect also wherever there is a slash (/) , an error is created

**Comment by weltkante** [comment]
in other words, the bug is in the code which builds the string, not in the code which parses the string

**Comment by ghost1372** [comment]
@miloush  tnx but your fix not worked and i get exception again

**Comment by weltkante** [comment]
Well, there are probably more places where the code is broken, not just that one. Every formatting of numbers needs to be done in the correct culture.

**Comment by ghost1372** [comment]
@weltkante  oh tnx Let me try again

**Comment by ghost1372** [comment]
@weltkante  @miloush  I used this code wherever possible but no results

**Comment by miloush** [comment]
Do you get the exception at the same place for the same reason?

**Comment by ghost1372** [comment]
@miloush  yes

**Comment by miloush** [comment]
How does the string look like now?

**Comment by ghost1372** [comment]
@miloush  same

`
System.FormatException: 'Unexpected token 'M41/416254,90C41/88876,82/5298 41/624584,57/2716 41/416254,50/99 36/29352,49/8907 36/894651,50/4264 31/567267,50/99 31/775597,44/59305 31/931627,42/43234 31/693536,36/77758 36/340008,36/79568 35/133525,36/76858 41/542523,36/58818 41/638383,32/81458 41/675605,33/50908 41/668792,29/20144 41/851382,25/46201 41/182183,18/65836 45/961941,14/23856 50/741699,9/81876 59/568752,10/59048 68/185297,10/70302 68/141127,16/44056 68/145937,20/8288 68/185297,23/58247 63/135373,24/4635 60/876109,23/24558 57/326157,24/84516 55/922779,28/231 55/756556,29/65047 56/63466,36/77758 59/21143,36/64422 64/265254,36/63978 68/690373,36/77758 67/695215,42/78657 67/742088,44/40134 66/922606,49/8464 63/245981,49/9344 61/316885,49/6976 55/810928,49/8464 55/662114,56/6024 56/168075,82/8571 55/810928,90 49/382352,90/1786 47/63649,89/97 41/416254,90z' encountered at position '2'.'
`

**Comment by miloush** [comment]
```
> 41.416254.ToString(CultureInfo.GetCultureInfo("fa-IR"))
"41/416254"
> 41.416254.ToString(CultureInfo.InvariantCulture)
"41.416254"
```

Should be working... are you sure you are running the updated code? Do you have the new code somewhere to check?

**Comment by ghost1372** [comment]
@miloush  yes i run updated code you can check my code [here](https://github.com/ghost1372/HandyControls/commit/f92406c938d636190f524e7dd9914b5eb0e61d78)

also i tested your code in this way
var xx = "41/416254";
            Debug.WriteLine(xx.ToString(CultureInfo.InvariantCulture));
            Debug.WriteLine(xx.ToString(CultureInfo.GetCultureInfo("fa-IR")));
            Debug.WriteLine(xx.ToString(new CultureInfo("en-US")));

and Neither worked and the result was the same (41/416254)

**Comment by weltkante** [comment]
> var xx = "41/416254";

thats not a number, thats a string (you are doing `string xx = "41/416254"` which you can verify via intellisense tooltips). The code you posted was using `double` (which is a number). Doubles need to be formatted with invariant culture when turned into strings for passing it on to WPF geometry parsing.

**Comment by ghost1372** [comment]
@weltkante  I'm a little confused
I can only change the double format right?
However, when we are creating geometries, the double format doesn't change according to the code @miloush wrote

```
internal static Geometry ComposeGeometry(string[] strings, double[] arr)
        {
            var builder = new StringBuilder(strings[0]);
            for (var i = 0; i < arr.Length; i++)
            {
                var s = strings[i + 1];
                var n = arr[i];
                if (!double.IsNaN(n))
                {
                    builder.Append(n.ToString(CultureInfo.InvariantCulture)).Append(s);
                }
            }
            return Geometry.Parse(builder.ToString());
        }
```

**Comment by miloush** [comment]
Okay, I had a look at the code, and it is a bit more complicated. As @weltkante said, `.ToString(CultureInfo.InvariantCulture)` has no effect on strings, so you don't need to add that to string variables. However, the double values in `arr` will actually always by integer numbers, because it's split in a bit crazy manner in the geometry animation:

https://github.com/ghost1372/HandyControls/blob/8afaf3f921be9990e8014038f4064c5a900807c6/src/Shared/HandyControl_Shared/Media/Animation/GeometryAnimationUsingKeyFrames.cs#L30

For some reason which I can see might be a bit controversial yet consistent with the framework, `Geometry.ToString()` uses current culture to format the string. In the light of the above, my previous advice does nothing to mitigate the problem, my apologies. The fix of the formatting problem you are having is to edit the above line to read `_keyFrames[0].Value.ToString(CultureInfo.InvariantCulture)`.

Note that I get an `IndexOutOfRangeException` anyway but that I guess is something for you to work on.

**Comment by ghost1372** [comment]
@miloush  Yes, I got an error like that `IndexOutOfRangeException`

Thank you very much for your help
